velox/functions/lib/aggregates/noisy_aggregation/sketch/tests/SfmSketchTest.cpp

```diff
@@ -0,0 +1,457 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/lib/aggregates/noisy_aggregation/sketch/SfmSketch.h"
+#include "gtest/gtest.h"
+#include "velox/common/encode/Base64.h"
+#include "velox/common/memory/Memory.h"
+#include "velox/functions/lib/aggregates/noisy_aggregation/sketch/MersenneTwisterRandomizationStrategy.h"
+
+namespace facebook::velox::functions::aggregate {
+
+class SfmSketchTest : public ::testing::Test {
+ public:
+  static void SetUpTestCase() {
+    memory::MemoryManager::testingSetInstance(memory::MemoryManager::Options{});
+  }
+
+ protected:
+  uint32_t numberOfBuckets_ = 4096;
+  uint32_t precision_ = 24;
+  std::shared_ptr<memory::MemoryPool> pool_{
+      memory::memoryManager()->addLeafPool()};
+  HashStringAllocator allocator_{pool_.get()};
+};
+
+TEST_F(SfmSketchTest, computeIndex) {
+  std::vector<uint32_t> indexBitLength = {6, 8, 10, 12};
+  for (auto& length : indexBitLength) {
+    uint32_t index = 5;
+    auto hash = static_cast<uint64_t>(index) << (64 - length);
+    ASSERT_EQ(SfmSketch::computeIndex(hash, length), index);
+  }
+}
+
+TEST_F(SfmSketchTest, numOfZeros) {
+  std::vector<uint32_t> indexBitLength = {6, 8, 10, 12};
+  for (auto& length : indexBitLength) {
+    for (uint32_t zeros = 0; zeros < 63; zeros++) {
+      uint64_t hash = 1ULL << zeros;
+      uint32_t kBitWidth = sizeof(uint64_t) * 8;
+      uint64_t value = hash | (1ULL << (kBitWidth - length));
+
+      auto trailingZeros = static_cast<uint32_t>(__builtin_ctzll(value));
+      ASSERT_EQ(trailingZeros, std::min(zeros, 64 - length));
+    }
+  }
+}
+
+TEST_F(SfmSketchTest, numOfbuckets) {
+  std::vector<uint32_t> indexBitLength = {6, 8, 10, 12};
+
+  for (auto& length : indexBitLength) {
+    ASSERT_EQ(SfmSketch::numberOfBuckets(length), 1U << length);
+  }
+}
+
+TEST_F(SfmSketchTest, bitMapSize) {
+  std::vector<uint32_t> numOfBuckets = {32, 64, 512, 1024, 4096, 32768};
+  std::vector<uint32_t> precisions = {1, 2, 3, 8, 24, 32};
+
+  auto pool = memory::memoryManager()->addLeafPool();
+  HashStringAllocator allocator(pool.get());
+
+  for (auto& numOfBucket : numOfBuckets) {
+    for (auto& precision : precisions) {
+      auto sketch = SfmSketch(&allocator);
+      sketch.initialize(numOfBucket, precision);
+      ASSERT_EQ(sketch.numberOfBits(), numOfBucket * precision);
+    }
+  }
+}
+
+TEST_F(SfmSketchTest, privacyEnabled) {
+  SfmSketch sketch(&allocator_);
+  sketch.initialize(numberOfBuckets_, precision_);
+  ASSERT_FALSE(sketch.privacyEnabled());
+  sketch.enablePrivacy(std::numeric_limits<double>::infinity());
+  ASSERT_FALSE(sketch.privacyEnabled());
+  sketch.enablePrivacy(1.0);
+  ASSERT_TRUE(sketch.privacyEnabled());
+}
+
+TEST_F(SfmSketchTest, mergeNonPrivacy) {
+  auto sketch1 = SfmSketch(&allocator_);
```

Comment:
a1, a2, a3 variable names are an anti-pattern

consider replacing sketch1 and sketch2 with a and b